text/0000-simple_unwind_annotation.md

```diff
@@ -0,0 +1,164 @@
+
+# Summary
+[summary]: #summary
+
+Provides a new ABI string `extern "C panic"` to denote functions that use the C ABI, but may also
+unwind with a Rust panic.
+
+# Motivation
+[motivation]: #motivation
+
+TODO
+
+- soundness & optimization
+- libjpeg/mozjpeg
+
+## Generated code
+
+When working with generated code, such as the result of a JIT compiler, calls to generated code
+frequently must use the C ABI. Similarly to the `mozjpeg` case, we want it to be possible for panics
+to unwind across calls to foreign code. However the dynamic nature of the foreign code distinguishes
+this case from the `mozjpeg` case, specifically because the set of foreign functions that may be
+called is not known at Rust compile time.
+
+This property makes wrappers that use `setjmp`/`longjmp` or C++ exceptions much more difficult to
+generate, as we have no header files to provide to a wrapper generator like
+[`ffi_wrapper_nounwind`][ffi_wrapper_nounwind]. It further means that a solution must accommodate
+foreign function pointers that don't have an `extern "C" { ... }` declaration where we can attach an
+attribute.
+
+[Lucet][lucet] and [Weld][weld] are two projects with these FFI use patterns that would concretely
+benefit from permitting unwinding through FFI boundaries.
+
+[ffi_wrapper_nounwind]: https://docs.rs/ffi_wrapper_nounwind
+[lucet]: https://github.com/fastly/lucet
+[weld]: https://www.weld.rs/
+
+## Less dependence on C/C++
+
+Some of the [proposed workarounds][cffi-panic] to the lack of cross-FFI unwinding require the use of
+wrappers written in C (for `setjmp` and `longjmp`) or C++ (for exceptions). This solution reduces
+the amount of non-Rust code that must be generated and maintained when the foreign code is
+compatible with the unspecified Rust unwinding mechanism.
+
+[cffi-panic]: https://github.com/gnzlbg/cffi-panic
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rust function definitions with the `"C panic"` ABI string (e.g., `extern "C panic" fn`) are
+permitted to unwind with a panic, as opposed to `extern "C" fn` functions which will abort the
+process if a panic reaches the function boundary.
+
+When used on declarations of imported functions (e.g., `extern "C panic" { fn ... }`) or function
+pointers (e.g., `extern "C panic" fn()`), the `"C panic"` ABI string means that if the function
+unwinds, the unwind will be propagated though any calling code. If an `extern "C"` imported function
+or function pointer unwinds, the behavior is undefined.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Currently, Rust assumes that an external function imported with `extern "C" {
+... }` (or another ABI other than `"Rust"`) cannot unwind, and Rust will abort if a panic would propagate out of a
+Rust function with a non-"Rust" ABI ("`extern "ABI" fn`") specification. Under this RFC,
+functions with the `"C panic"` ABI string
```

Comment:
`extern "system panic"` may also be required to replace current uses of `#[unwind(allow)]`, here's an example (WinAPI used in libstd):

https://github.com/rust-lang/rust/blob/master/src/libpanic_unwind/windows.rs#L71-L86